### ISSUE TITLE
Does this Polyfill 1-to-1?

### ISSUE THREAD

**Issue by KooIaIa** [WebReflection/sabayon#4]
If I understand right, if I include this as a script in my main site namespace, my workers, and my service worker - can I use SharedArrayBuffers like normal without any code differences?

**Comment by WebReflection** [comment]
does [this part of the README](https://github.com/WebReflection/sabayon?tab=readme-ov-file#caveats) answer?

*TL;DR*

  * if you can't serve headers on the server, you need a ServiceWorker helper
    * if that SW helper is `mini-coi` you have feature parity
    * if that SW is the `sabayon` one or your own one with `sabayon` helpers integration, you have *XMLHttpRequest* emulation which works almost identically but it's a bit slower (see performance section) and has the mentioned caveats
  * if you can't have a ServiceWorker and/or right headers, you still have the asynchronous polyfill working as expected **but**:
    * you cannot wait synchronously
    * you won't have *count* or *interrupts* as specified in the caveats
  
I hope this answers your question.

**Comment by KooIaIa** [comment]
I saw your blog qbout a service worker polyfill. A way to use service worker code on any site while ignoring cors would be amazing.

I was trying to see if the polyfill wouldn't create new JS semantics. So service worker code could be written without the sabayon library and work with headers - and then work with the sabayon library without headers.

mini-coi sounded like a good option as a way to set headers when you can't set headers but then it still needs headers to work?

"always require-corp to have this working on Safari as well as Chrome or Firefox"

**Comment by WebReflection** [comment]
I use [mini-coi sw](https://github.com/WebReflection/mini-coi?tab=readme-ov-file#service-worker) on GitHub pages and that's it ... with that you don't even need sabayon but if you have it, nothing changes, you are on native *SharedArrayBuffer* there.

Sabayon is for cases where:

  * you need SharedArrayBufer async no matter what (can't use COI headers, can't use SW, can't change headers)
  * you need synchronous SharedArrayBuffer (you already have a SW so you can add sabayon helpers or you want sync Atomics without the burden that COI headers require, still via the sabayon Service Worker)
  * you need synchronous SharedArrayBuffer and you just want to be sure it works ... in this case, if headers or a Service Worker that allows COI is in place, you just don't worry about it as it won't need to polyfill anything ... this is the code-portability option

**Comment by WebReflection** [comment]
A production example: PyScript uses sabayon which does nothing on pyscript.com but it's portable on GitHub pages or any other service that can add mini-coi or can add sabayon helpers to the currently running service worker https://docs.pyscript.net/2024.10.1/user-guide/workers/
